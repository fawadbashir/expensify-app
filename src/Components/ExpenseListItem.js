import React from 'react'
import moment from 'moment'
import numeral from 'numeral'
import { Link } from 'react-router-dom'
import Card from 'react-bootstrap/Card'



const ExpenseListItem = ({ id, description, amount, createdAt }) => (


    <Card.Body style={{ borderBottom: '1px solid #e5e5e5' }} >

        <Link className='list-group-item-action' to={`/edit/${id}`}>
            <Card.Title bg='dark' as='h3'>{description}</Card.Title>
        </Link>



        <Card.Subtitle className='mt-4'> {moment(createdAt).format('MMMM do YYYY')}
            <Card.Text as='span' className='float-right '>{numeral(amount / 100).format('$0,0.00')}</Card.Text>
        </Card.Subtitle>



    </Card.Body>





)
export default ExpenseListItem