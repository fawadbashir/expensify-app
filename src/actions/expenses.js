import database from '../firebase/firebase'

//ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return async (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }

    const ref = await database.ref('expenses').push(expense)
    dispatch(addExpense({
      id: ref.key,
      ...expense
    }))
  }
}

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
})

 const startRemoveExpense = ({id} = {}) => {
  return async (dispatch) => { 
 await database.ref(`expenses/${id}`).remove()
     dispatch(removeExpense({id}))
  }

}

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
})

//SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return  async (dispatch) => {
   const snapshot = await database.ref('expenses').once('value')
    const expenses = []
    snapshot.forEach((childSnapshot) => {
      expenses.push({
        id: childSnapshot.key,
        ...childSnapshot.val()
      })
    })
    dispatch(setExpenses(expenses))
  }
}

export { startRemoveExpense }