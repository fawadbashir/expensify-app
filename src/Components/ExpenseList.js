import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

const ExpenseList =  (props) =>{
    return <div>
        <h1>ExpenseList</h1> 
            {/* <p>{props.expenses.length}</p> */}
            {props.expenses.map((expense)=> 
            <ExpenseListItem key = {expense.id} {...expense}/>)}
    </div>
}


const mapStatetoProps = (state) => {
    return {
        expenses : selectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStatetoProps)(ExpenseList)