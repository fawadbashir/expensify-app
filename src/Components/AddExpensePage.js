import React, { Component } from 'react';
import Typography from '@material-ui/core/Typography'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense)
        this.props.history.push('/dashboard')
    }

    render() {
        return (
            <div>
            {/* <Typography variant="h3" color="blue">Some text in here</Typography> */}
               
                <Typography variant = 'h3'>Add Expense</Typography>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
                
            </div>
        )
    }
}



const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage)