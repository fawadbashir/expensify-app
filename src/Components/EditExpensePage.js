import React, { Component } from 'react';
import { connect } from 'react-redux'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends Component {
    onSubmit = (expense)=>{
        this.props.startEditExpense(this.props.expense.id,expense)
        this.props.history.push('/')
    }

    onClick = () => {
        this.props.startRemoveExpense({ id : this.props.expense.id})
        
        this.props.history.push('/')    
    }
    render() {
        return (
            <div>
        <ExpenseForm
        expense = {this.props.expense}
            onSubmit = {this.onSubmit}
        />
        <button onClick = {this.onClick}>Remove</button>
    </div>
        )
    }
}

// const EditExpensePage = (props) =>{
    
//     return (
//         <div>
//         <ExpenseForm
//         expense = {props.expense}
//             onSubmit = {(expense)=>{
//                 props.dispatch(editExpense(props.expense.id,expense))
//                 props.history.push('/')
//             }}
//         />
//         <button onClick = {() => {
//             props.dispatch(removeExpense({ id : props.expense.id})
//             )
//             props.history.push('/')    
//         }}
            
//             >Remove</button>
//     </div>
//     )
// }
const mapStateToProps = (state,props) => {
    return {
        expense : state.expenses.find((expense)=>expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => 
    ({startEditExpense: (expense) => dispatch(startEditExpense(expense)),
        startRemoveExpense: (expense) => dispatch(startRemoveExpense({id : expense.id}))})


export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)