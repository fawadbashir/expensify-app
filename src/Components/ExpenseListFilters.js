import React, { Component } from 'react'
import { connect } from 'react-redux'
import {DateRangePicker} from 'react-dates'
import { setTextFilter, sortByDate, sortByAmount,setStartDate,setEndDate } from '../actions/filters'
import uuidv4 from 'uuidv4'

class expenseListFilters extends Component {
    state = {
        focusedInput : null
    }
    onDatesChange =({startDate,endDate})=>{
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }
    onFocusChange = (focusedInput)=>{this.setState(()=>({focusedInput}))}

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e) => {
                   this.props.dispatch(setTextFilter(e.target.value))
                }} />
                <select value={this.props.filters.sortBy} onChange={(e) => {
                    if (e.target.value === 'date') {
                        this.props.dispatch(sortByDate())
                    } else {
                        this.props.dispatch(sortByAmount())
                    }
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                startDate = {this.props.filters.startDate}
                startDateId = {uuidv4()}
                endDate = {this.props.filters.endDate}
                endDateId= {uuidv4()}
                onDatesChange = {this.onDatesChange}
                focusedInput ={this.state.focusedInput}
                onFocusChange = {this.onFocusChange}
                isOutsideRange ={()=> false}

                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
export default connect(mapStateToProps)(expenseListFilters)