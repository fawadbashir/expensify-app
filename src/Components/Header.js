import React from 'react';
import { connect } from 'react-redux'
import {NavLink} from 'react-router-dom'
import { startLogout } from '../actions/auth'

 export const Header = (props) => {
     console.log(props)
     return(
        <header>
    <h1>Expensify</h1>   
     <NavLink activeClassName = 'active' to = '/'>Home</NavLink>      
    <NavLink activeClassName = 'active' to = '/create'>Add Expense </NavLink>   
    <NavLink  activeClassName = 'active' to = '/help'>Help</NavLink>   
    <button onClick = {props.startLogout}>LogOut</button>
    </header>
     )
     }
    


const mapDispatchToProps = (dispatch) => ({
    startLogout : () => {
        return dispatch(startLogout());
    }
})

    


export default connect( mapDispatchToProps)(Header)