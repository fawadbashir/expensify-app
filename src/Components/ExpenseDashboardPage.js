import React from 'react';
import { Link } from 'react-router-dom'
import Button from 'react-bootstrap/Button'


import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
    <div>
        <div className='pt-3 full-width-body' style={{ marginBottom: '0.5rem', padding: '0.5rem 0', }}>
            <ExpensesSummary />
            <Link to='/create'><Button size='lg' className=' mt-3' style={{ fontSize: '18px', backgroundColor: '#3D9970', color: 'white', border: 'none', padding: '10px' }} >Add Expense</Button></Link>
        </div>
        <ExpenseListFilters />
        <ExpenseList />



    </div >
)

export default ExpenseDashboardPage