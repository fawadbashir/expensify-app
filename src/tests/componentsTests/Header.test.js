import React from 'react'
import {createSerializer} from 'enzyme-to-json';
import { shallow } from 'enzyme'
import {Header} from '../../Components/Header'
import { startLogout } from '../../actions/auth'
expect.addSnapshotSerializer(createSerializer({mode: 'shallow'}));




test('should render Header correctly', () => {
    

    const wrapper = shallow(<Header startLogout = {startLogout}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should call startLogout on button click', () => {
    const startLogout= jest.fn()
    const wrapper = shallow(<Header startLogout = {startLogout}/>)
    wrapper.find('button').simulate('click')
    expect(startLogout).toHaveBeenCalled()
})