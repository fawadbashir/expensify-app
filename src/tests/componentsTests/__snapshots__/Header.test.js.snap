// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="active"
    aria-current="page"
    to="/"
  >
    Home
  </NavLink>
  <NavLink
    activeClassName="active"
    aria-current="page"
    to="/create"
  >
    Add Expense 
  </NavLink>
  <NavLink
    activeClassName="active"
    aria-current="page"
    to="/help"
  >
    Help
  </NavLink>
  <button
    onClick={[Function]}
  >
    LogOut
  </button>
</header>
`;
